### DISCUSSION TITLE
EMC++ 항목 17에 관련된 질문

### DISCUSSION THREAD

**Discussion by Lcat12345** [MyeongWoonJang/cpp-study#155] [Q&A] [answered]
항목 17을 읽던 도중 이런식으로 이해하였습니다.

> - 이동 생성자를 작성하게 되면 컴파일러가 이동 대입 연산자, 복사 생성자, 복사 대입 연산자를 작성하지 못하게 한다.
이동 대입연산자도 마찬가지로 나머지 이동 연산과 복사연산을 작성하지 못하게 한다(delete하여).
>- 그런데 복사와 이동 연산에 default를 붙여 선언하면 복사와 이동 연산을 동시에 지원한다.
>- 템플릿 멤버 함수는 예외다.

그런데 작성하지 못하게 한다로만 이해하니 헷갈렸습니다.(내가 못쓰는겨 컴파일러가 못쓴다는 겨)

사용자가 직접 이동 연산(또는 복사 연산)을 선언하였다. 
-> 
**"복사 연산(또는 이동 연산)을 금지(delete) 했으니 선언 할수 없다"** 
라는 뜻 일까요? 아니면 
**"복사 또는 이동 연산을 선언 한 걸 보니,  나머지를 자동으로 작성해주는 것은 옳지 않아보여서 자동으로 작성해주는 것을 금지시키는 거고 사용자가 직접 선언하라" 라는 뜻인가요?**

만일 후자의 경우면 default를 붙인 이유가 내가 복사 또는 이동 연산을 선언 하지 않을것이니 컴파일러가 자동으로 쓰게 해주는 것일까요?

**Comment by Woon-2** [comment, marked as answer]
후자입니다.   

`(special member funtion declaration) = default;`와 같은 구문에는 이미 `declaration`이 포함되어 있습니다.   
`= default`의 의미는 그 `definition`을 `compiler`가 제공하라고 요구하는 것이죠.

항목 17의 제목이 "특별 멤버 함수가 언제 자동으로 작성되는지 숙지하라."   
이런 느낌일 텐데,   

제목을 다시 보시면 '아~ 여기서 다루는 내용은 다 자동 작성에 관한 얘기구나' 하실 겁니다.

[\[cppreference\] The rule of three/five/zero](https://en.cppreference.com/w/cpp/language/rule_of_three)도 한번 읽어보세요.   

질문하신 내용과 관련해서 `compiler`의 그러한 행동들의 원인은 구닥다리 `rule of three`의 잔재이고,   
현재는 `rule of zero`가 추구된다는 걸 이해하세요.   

`special member function`을 직접 구현한다면 5개 전부 직접 구현하고,   
그렇지 않을 경우 5개 전부 `compiler`에게 맡기라는 규칙입니다.

## 멤버 함수 템플릿에 대한 보충 설명

> - 템플릿 멤버 함수는 예외다.

이렇게 적어놓으신 걸 보니, 완전히 이해하고 넘어가시는 것 같지 않습니다.

저는 `Scott Meyers`와 달리,   
`function template(함수 템플릿)`과 `template function(템플릿 함수)`을 구분합니다.   

- `function template`은 `function`을 찍어내는 틀이지, `function`이 아닙니다.
- `template function`은 `a function instantiated  from a function template`을 나타냅니다.   
  (함수 템플릿으로부터 인스턴스화된 함수)

```cpp
class Foo { ... };

// function template
template <class T>
void bar() { ... };

...

int main()
{
    bar<Foo>();    // call template function
   // = call function instantiated from function template
}
```

`non-template function`의 경우 여러분이 `function`을 전부 구현해놓고 해당 `function`을 이용하지만, 
`function template`의 경우 여러분이 `function`의 이용 사례들을 먼저 적어놓고 `compiler`에게 구현을 부탁하는 겁니다.

`explicit instantiation(명시적 인스턴스화)`나 `template function call`이 `code`에 존재하지 않을 경우,   
`function template`은 아무런 `code`도 구현해내지 않습니다.   

마치 그런 `function template`이 존재하지 않았던 것처럼요.

<br>

이런 사실들을 숙지하셨다면,   
다음으로 `code`의 `compile` 과정을 봅시다.   

자, 질문자분께서 어떠한 `class` `Foo`를 `define` 했다고 생각해봅시다.

그 과정 속에서 `special member function`을 대체할 만한 `member function template`을 추가했습니다.

그리고 `Foo` 및 `Foo`를 이용하는 `code`들을 완성했습니다.    

좋습니다.   
질문자분께서는 이제 `compiler`를 통해 `code`를 `compile`합니다.

`compiler`는 `Foo`를 이용하는 `code`보다는 `Foo`를 먼저 `compile`합니다.   
당연한 이치죠,   
만들어지지도 않은 것을 사용할 순 없잖아요?

`compiler`는 `Foo`를 `compile`하면서,   
**아직까지 `member function template`이 한 번도 `instantiate`된 적이 없기 때문에**   
**그 어떠한 `special member function`도 발견하지 못했습니다.**

따라서 `compiler`는 `special member function`을 직접 작성하죠.

그렇게 `Foo`가 `compile`되었습니다.   
이제 `Foo`를 이용하는 `code`들도 `compile`할 수 있겠죠.

그리고 그 `code`에는 `special member function`을 이용하는 문맥이 있습니다.

문제는, `compiler`가 만든 `special member function`도 이용할 수 있고,   
`member function template`도 이용할 수 있어서 둘 중 하나를 선택해야 한다는 것입니다.

`compiler`는 두 가지 결정을 할 수 있습니다.

- 이미 만들어놓은 `special member function`을 쓰기
- `member function template`을 `instantiate`하기

`parameter`가 확실하게 부합하는 경우,   
`compiler`는 새롭게 `function`을 만들어 놓는 것보다 기존의 `function`을 쓰는 게 비용이 적다고 판단하기 때문에
이미 만들어놓은 `special member function`을 쓰는 것을 택합니다.

"`parameter`가 확실하게 부합하는" `special member function call`만 있었다면,   
`template function`이 불릴 것이라 기대했던 여러분의 예상과는 달리   
`compiler`가 만든 `special member function`이 불릴 겁니다. 

"확실하게 부합하는"을 따로 강조하는 이유는,
EMC++의 항목 26을 보면 아실 겁니다.
